### PR TITLE
fix(parallel-execution): Adding parallel execution in pod cpu/memory-hog experiments

### PR DESCRIPTION
**Pull request by ispeakc0de**
Signed-off-by: shubhamchaudhary <shubham.chaudhary@mayadata.io>

- Adding parellel execution in pod cpu/memory-hog experiments

### CODE REVIEW

**Comment by ispeakc0de** [comment]
## abort case in parellel execution 

```yaml
time="2020-09-24T07:39:54Z" level=info msg="[PreReq]: Getting the ENV for the  experiment"
time="2020-09-24T07:39:54Z" level=info msg="[PreReq]: Updating the chaos result of pod-memory-hog experiment (SOT)"
time="2020-09-24T07:39:54Z" level=info msg="The application information is as follows" Memory Consumption=500 Namespace=shubham Label="run=nginx" Chaos Duration=100 Ramp Time=0
time="2020-09-24T07:39:54Z" level=info msg="[Status]: Verify that the AUT (Application Under Test) is running (pre-chaos)"
time="2020-09-24T07:39:54Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-09-24T07:39:54Z" level=info msg="The running status of container are as follows" Pod=nginx-6c444d977f-cv6sh Status=Running container=nginx
time="2020-09-24T07:39:54Z" level=info msg="The running status of container are as follows" container=nginx Pod=nginx-6c444d977f-r4qpm Status=Running
time="2020-09-24T07:39:54Z" level=info msg="The running status of container are as follows" container=nginx Pod=nginx-6c444d977f-wkz9f Status=Running
time="2020-09-24T07:39:56Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-09-24T07:39:56Z" level=info msg="The running status of Pods are as follows" Pod=nginx-6c444d977f-cv6sh Status=Running
time="2020-09-24T07:39:56Z" level=info msg="The running status of Pods are as follows" Status=Running Pod=nginx-6c444d977f-r4qpm
time="2020-09-24T07:39:56Z" level=info msg="The running status of Pods are as follows" Status=Running Pod=nginx-6c444d977f-wkz9f
time="2020-09-24T07:39:58Z" level=info msg="[Chaos]:Number of pods targeted: 3"
time="2020-09-24T07:39:58Z" level=info msg="[Chaos]: The Target application details" Memory Consumption(MB)=500 container=nginx Pod=nginx-6c444d977f-wkz9f
time="2020-09-24T07:39:58Z" level=info msg="The memory consumption is: 500"
time="2020-09-24T07:39:58Z" level=info msg="[Chaos]: The Target application details" container=nginx Pod=nginx-6c444d977f-cv6sh Memory Consumption(MB)=500
time="2020-09-24T07:39:58Z" level=info msg="The memory consumption is: 500"
time="2020-09-24T07:39:58Z" level=info msg="[Chaos]: The Target application details" container=nginx Pod=nginx-6c444d977f-r4qpm Memory Consumption(MB)=500
time="2020-09-24T07:39:58Z" level=info msg="[Chaos]:Waiting for: 100s"
time="2020-09-24T07:39:58Z" level=info msg="The memory consumption is: 500"
time="2020-09-24T07:40:38Z" level=info msg="[Chaos]: Killing process started because of terminated signal received"
Terminated
Terminated
Terminated

```

**Comment by ksatchit** [comment]
@andreas131989 , would love to hear your opinions on these changes. 

Context: We are trying to introduce the ability to perform parallel chaos (control blast radius via the POD_AFFECT_PERC environment variable) on all the pod-level experiments. As part of this, the pod-cpu & memory hog experiments are being refactored to pick the target container and induce the resource hog process on one or more of the target pods at the same time (over inducing the process on "all" containers of one pod at a time, in sequence). Do you think this needs to be a separate lib or be controlled via some *mode*? OR the changes are good to be taken as is.

**Comment by andreas131989** [comment]
> @andreas131989 , would love to hear your opinions on these changes.
> 
> Context: We are trying to introduce the ability to perform parallel chaos (control blast radius via the POD_AFFECT_PERC environment variable) on all the pod-level experiments. As part of this, the pod-cpu & memory hog experiments are being refactored to pick the target container and induce the resource hog process on one or more of the target pods at the same time (over inducing the process on "all" containers of one pod at a time, in sequence). Do you think this needs to be a separate lib or be controlled via some _mode_? OR the changes are good to be taken as is.

Hey Karthik!
This is an improvement that we also just did a couple of days ago as well. The approach we are taking for such use cases, is with a certain mode. I think there is no need for a separate library as it's really a configuration matter. For example, we are thinking to use an env variable called ''sequence'' that can take values serial/parallel and then the runner performs the experiment accordingly.

File: pkg/generic/pod-memory-hog/environment/environment.go [hunk 1]
```diff
@@ -31,6 +31,8 @@ func GetENV(experimentDetails *experimentTypes.ExperimentDetails) {
 	experimentDetails.TargetPod = Getenv("TARGET_POD", "")
 	experimentDetails.ChaosKillCmd = Getenv("CHAOS_KILL_COMMAND", "kill $(find /proc -name exe -lname '*/dd' 2>&1 | grep -v 'Permission denied' | awk -F/ '{print $(NF-1)}' |  head -n 1)")
 	experimentDetails.LIBImage = Getenv("LIB_IMAGE", "gaiaadm/pumba")
+	experimentDetails.TargetContainer = Getenv("TARGET_CONTAINER", "")
+	experimentDetails.Sequence = Getenv("SEQUENCE", "parellel")
```
Comment (ksatchit) [inline pkg/generic/pod-memory-hog/environment/environment.go:35]:
```suggestion
	experimentDetails.Sequence = Getenv("SEQUENCE", "parallel")
```

File: pkg/generic/pod-cpu-hog/environment/environment.go [hunk 2]
```diff
@@ -32,6 +32,8 @@ func GetENV(experimentDetails *experimentTypes.ExperimentDetails) {
 	experimentDetails.ChaosInjectCmd = Getenv("CHAOS_INJECT_COMMAND", "md5sum /dev/zero")
 	experimentDetails.ChaosKillCmd = Getenv("CHAOS_KILL_COMMAND", "kill $(find /proc -name exe -lname '*/md5sum' 2>&1 | grep -v 'Permission denied' | awk -F/ '{print $(NF-1)}' |  head -n 1)")
 	experimentDetails.LIBImage = Getenv("LIB_IMAGE", "gaiaadm/pumba")
+	experimentDetails.TargetContainer = Getenv("TARGET_CONTAINER", "")
+	experimentDetails.Sequence = Getenv("SEQUENCE", "parellel")
```
Comment (ksatchit) [inline pkg/generic/pod-cpu-hog/environment/environment.go:36]:
```suggestion
	experimentDetails.Sequence = Getenv("SEQUENCE", "parallel")
```

**Review by ksatchit — APPROVED** [review]

**Pull request merged**